Retention rules API documentation refactor (#14623)
apache/druid: docs/api-reference/retention-rules-api.md

  ~ under the License.
  -->

This topic describes the API endpoints for managing retention rules in Apache Druid. You can configure retention rules in the Druid web console or API.

> Druid uses retention rules to determine what data is retained in the cluster. Druid supports load, drop, and broadcast rules. See [using rules to drop and retain data](../operations/rule-configuration.md) for more information. 

Reviewer: ```suggestion
Druid uses retention rules to determine what data is retained in the cluster. Druid supports load, drop, and broadcast rules. See [Using rules to drop and retain data](../operations/rule-configuration.md) for more information. 
```

In this topic, `http://ROUTER_IP:ROUTER_PORT` is a placeholder for your Router service address and port. Replace it with the information for your deployment. For example, use `http://localhost:8888` for quickstart deployments.

## Update retention rules for a datasource

> Update one or more retention rules for a datasource. Retention rules can be submitted as an array of rule objects in the request body and overwrite any existing rules for the datasource. Rules are read in the order in which they appear, see [rule structure](../operations/rule-configuration.md) for more information.

Reviewer: Active voice, and link to reduce friction for creating the request
```suggestion
Updates one or more retention rules for a datasource. The request body takes an array of retention rule objects. For details on defining retention rules, see the following sources:
* [Load rules](../operations/rule-configuration.md#load-rules)
* [Drop rules](../operations/rule-configuration.md#drop-rules)
* [Broadcast rules](../operations/rule-configuration.md#broadcast-rules)

This request overwrites any existing rules for the datasource.

Druid reads rules in the order in which they appear; for more information, see [rule structure](../operations/rule-configuration.md).
```

<!--cURL-->

```shell
curl "http://ROUTER_IP:ROUTER_PORT/druid/coordinator/v1/rules/kttm1" \
--header 'X-Druid-Author: doc intern' \
--header 'X-Druid-Comment: submitted via api' \
--header 'Content-Type: application/json' \
--data '[
    {
        "type": "broadcastForever"
    },
    {
        "type": "loadForever",
        "tieredReplicants": {
            "_default_tier": 2
        },
        "useDefaultTierForNull": true
    },
    {
        "type": "dropByPeriod",
        "period": "P1M"
    }
]'
```

<!--HTTP-->

```HTTP
POST /druid/coordinator/v1/rules/kttm1 HTTP/1.1
Host: http://ROUTER_IP:ROUTER_PORT
X-Druid-Author: doc intern
X-Druid-Comment: submitted via api
Content-Type: application/json
Content-Length: 273

[
    {
        "type": "broadcastForever"
    },
    {
        "type": "loadForever",
        "tieredReplicants": {
            "_default_tier": 1
        },
        "useDefaultTierForNull": true
    },
    {
        "type": "dropByPeriod",
        "period": "P1M"
    }
]
```

<!--END_DOCUSAURUS_CODE_TABS-->

### Sample response

A successful request returns an HTTP `200 OK` and an empty response body.

## Update default retention rules for all datasources

> Update one or more default retention rules for all datasources. Retention rules can be submitted as an array of rule objects in the request body and overwrite any existing rules for the datasource. To remove default retention rules for all datasources, submit an empty rule array in the request body. Rules are read in the order in which they appear, see [rule structure](../operations/rule-configuration.md) for more information.

Reviewer: ```suggestion
Updates one or more default retention rules for all datasources. Submit retention rules as an array of objects in the request body. This request overwrites any existing rules for all datasources. To remove default retention rules for all datasources, submit an empty rule array in the request body. Rules are read in the order in which they appear; for more information, see [rule structure](../operations/rule-configuration.md).
```

### Responses

<!--DOCUSAURUS_CODE_TABS-->

<!--200 SUCCESS-->

*Successfully updated default retention rules* 

<!--500 SERVER ERROR-->

*Error with request body* 

<!--END_DOCUSAURUS_CODE_TABS-->

---

### Sample request

The following example updates the default retention rule for all datasources with a `loadByInterval` rule.

<!--DOCUSAURUS_CODE_TABS-->

<!--cURL-->

```shell
curl "http://ROUTER_IP:ROUTER_PORT/druid/coordinator/v1/rules/_default" \
--header 'Content-Type: application/json' \
--data '[
    {
        "type": "loadByInterval",
        "tieredReplicants": {},
        "useDefaultTierForNull": false,
        "interval": "2010-01-01/2020-01-01"
    }
]'
```

<!--HTTP-->

```HTTP
POST /druid/coordinator/v1/rules/_default HTTP/1.1
Host: http://ROUTER_IP:ROUTER_PORT
Content-Type: application/json
Content-Length: 205

[
    {
        "type": "loadByInterval",
        "tieredReplicants": {},
        "useDefaultTierForNull": false,
        "interval": "2010-01-01/2020-01-01"
    }
]
```

<!--END_DOCUSAURUS_CODE_TABS-->

### Sample response

A successful request returns an HTTP `200 OK` and an empty response body.

## Get an array of all retention rules

> Retrieves all current retention rules in the cluster including the default retention rule. Returns an array of objects for each datasource and their associated retention rule.

Reviewer: ```suggestion
Retrieves all current retention rules in the cluster including the default retention rule. Returns an array of objects for each datasource and their associated retention rules.
```

### Responses

<!--DOCUSAURUS_CODE_TABS-->

<!--200 SUCCESS-->

*Successfully retrieved retention rules* 

<!--END_DOCUSAURUS_CODE_TABS-->

---

### Sample request

The following example retrieves the custom retention rules and default retention rules for datasource with the name `social_media`.

<!--DOCUSAURUS_CODE_TABS-->

<!--cURL-->

```shell
curl "http://ROUTER_IP:ROUTER_PORT/druid/coordinator/v1/rules/social_media?full=null"
```

<!--HTTP-->

```HTTP
GET /druid/coordinator/v1/rules/social_media?full=null HTTP/1.1
Host: http://ROUTER_IP:ROUTER_PORT
```

<!--END_DOCUSAURUS_CODE_TABS-->

### Sample response

<details>
  <summary>Click to show sample response</summary>

  ```json
[
    {
        "interval": "2020-01-01T00:00:00.000Z/2022-02-01T00:00:00.000Z",
        "type": "dropByInterval"
    },
    {
        "interval": "2010-01-01T00:00:00.000Z/2020-01-01T00:00:00.000Z",
        "tieredReplicants": {
            "_default_tier": 2
        },
        "type": "loadByInterval"
    },
    {
        "tieredReplicants": {
            "_default_tier": 2
        },
        "type": "loadForever"
    }
]
  ```

</details>

## Get audit history for all datasources

> Retrieves the audit history of rules for all datasources over an interval of time. The default value of `interval` can be specified by setting `druid.audit.manager.auditHistoryMillis` (1 week if not configured) in the `runtime.properties` file for the Coordinator.

Reviewer: ```suggestion
Retrieves the audit history of rules for all datasources over an interval of time. The default interval is 1 week. You can change this period by setting `druid.audit.manager.auditHistoryMillis` in the `runtime.properties` file for the Coordinator.
```

> This document describes the API endpoints for managing retention rules in Apache Druid.

Author: Wdym?

> This document describes the API endpoints for managing retention rules in Apache Druid.

Reviewer 2: N/m, there is a section on the retention rules page. Basically, I was asking if this existed: https://druid.apache.org/docs/latest/operations/rule-configuration.html#set-retention-rules

> This document describes the API endpoints for managing retention rules in Apache Druid.

Author: I added that you can do it in the console or API, but i didn't link it since the same destination is already linked in the sentence under it.

Druid uses retention rules to determine what data is retained in the cluster. Druid supports load, drop, and broadcast rules. See [using rules to drop and retain data](../operations/rule-configuration.md) for more information. 

> In this document, `http://SERVICE_IP:SERVICE_PORT` is a placeholder for the server address of deployment and the service port. For example, on the quickstart configuration, replace `http://ROUTER_IP:ROUTER_PORT` with `http://localhost:8888`.

Reviewer 2: I don't really understand the need for the double placeholder explanation when the router one is already a placeholder?

What does it do that the following doesn't?

`http://ROUTER_IP:ROUTER_PORT` is a place holder for your Router service address and port. Replace it with the information for your deployment. For example, use `http://localhost:8888` for quickstart deployments.

## Update retention rules for a datasource

Update one or more retention rules for a datasource. Retention rules can be submitted as an array of rule objects in the request body and overwrite any existing rules for the datasource. Rules are read in the order in which they appear, see [rule structure](../operations/rule-configuration.md) for more information.

> Note that this endpoint returns an `HTTP 200 Success` code message even if the `datasource` does not exist.

Reviewer 2: Why is this occurence of datasource in code font?

### URL
<code class="postAPI">POST</code> `/druid/coordinator/v1/rules/{datasource}`

### Header parameters

The endpoint supports a set of optional header parameters to populate the `author` and `comment` fields in the `auditInfo` property for audit history. 

* `X-Druid-Author` (optional)
  * Type: String
  * A string representing the author making the configuration change.
* `X-Druid-Comment` (optional)
  * Type: String
  * A string describing the update.

### Responses

<!--DOCUSAURUS_CODE_TABS-->

<!--200 SUCCESS-->
<br/>
*Successfully updated retention rules for specified datasource* 

<!--END_DOCUSAURUS_CODE_TABS-->

---

### Sample request

> The following example updates the retention rules for a datasource with specified ID `kttm1`.

Reviewer 2: Do we actually call it an ID? I thought it was a name? Most of the IDs we talk about in Druid are of the autogenerated sort I think, so it'd be nice to have that consistency unless it is actually called ID.

Same q for all the other instnaces of this.

### URL

<code class="getAPI">GET</code> `/druid/coordinator/v1/rules`

### Responses

<!--DOCUSAURUS_CODE_TABS-->

<!--200 SUCCESS-->
<br/>
*Successfully retrieved retention rules* 

<!--END_DOCUSAURUS_CODE_TABS-->

---

### Sample request

<!--DOCUSAURUS_CODE_TABS-->

<!--cURL-->
```shell
curl "http://ROUTER_IP:ROUTER_PORT/druid/coordinator/v1/rules"
```
<!--HTTP-->
```HTTP
GET /druid/coordinator/v1/rules HTTP/1.1
Host: http://ROUTER_IP:ROUTER_PORT
```
<!--END_DOCUSAURUS_CODE_TABS-->

### Sample response

<details>
  <summary>Click to show sample response</summary>

  ```json
{
    "_default": [
        {
            "tieredReplicants": {
                "_default_tier": 2
            },
            "type": "loadForever"
        }
    ],
    "social_media": [
        {
            "interval": "2023-01-01T00:00:00.000Z/2023-02-01T00:00:00.000Z",
            "type": "dropByInterval"
        }
    ],
    "wikipedia_api": [],
}
  ```
</details>

## Get an array of retention rules for a datasource

Retrieves an array of rule objects for a single datasource. If there are no retention rules, it returns an empty array. 

> Note that this endpoint returns an `HTTP 200 Success` code message even if the `datasource` does not exist.

Reviewer 2: Same comment above about this probably not needing to be in code font.

### Responses

<!--DOCUSAURUS_CODE_TABS-->

<!--200 SUCCESS-->
<br/>
*Successfully retrieved retention rules* 

<!--END_DOCUSAURUS_CODE_TABS-->

---

### Sample request

The following example retrieves the custom retention rules and default retention rules for datasource with specified ID `social_media`.

<!--DOCUSAURUS_CODE_TABS-->

<!--cURL-->
```shell
curl "http://ROUTER_IP:ROUTER_PORT/druid/coordinator/v1/rules/social_media?full=null"
```
<!--HTTP-->
```HTTP
GET /druid/coordinator/v1/rules/social_media?full=null HTTP/1.1
Host: http://ROUTER_IP:ROUTER_PORT
```
<!--END_DOCUSAURUS_CODE_TABS-->

### Sample response

<details>
  <summary>Click to show sample response</summary>

  ```json
[
    {
        "interval": "2020-01-01T00:00:00.000Z/2022-02-01T00:00:00.000Z",
        "type": "dropByInterval"
    },
    {
        "interval": "2010-01-01T00:00:00.000Z/2020-01-01T00:00:00.000Z",
        "tieredReplicants": {
            "_default_tier": 2
        },
        "type": "loadByInterval"
    },
    {
        "tieredReplicants": {
            "_default_tier": 2
        },
        "type": "loadForever"
    }
]
  ```
</details>

## Get audit history for all datasources

> Retrieves the audit history of rules for all datasources over an interval of time. The default value of `interval` can be specified by setting `druid.audit.manager.auditHistoryMillis` (1 week if not configured) in Coordinator `runtime.properties`.

Reviewer 2: Id'd rephrase this since the naming/file structure is a little unintuitive for Druid since all the runtime properties are just called `runtime.properties` regardless of service besides common and you have to check the directory

- in the runtime properties for the Coordinator 
- in the `runtime.properties` file for the Coordinator
